Add GDExtension support for OpenXR extension wrappers

konczg opened the pull request:
This commit adds the classes `OpenXRExtensionWrapperExtension` and `OpenXRAPIExtension` that can be used in GDExtensions to define OpenXR extension wrappers. It modifies extension wrapper registration so that they can be registered before `OpenXRAPI` instantiation (e.g. in core level initialization of GDExtensions).

As `OpenXRAPI` requires all extension wrappers to be registered upon initialization, the construction order of `OpenXRAPI` and the extension wrappers has been swapped, and extension wrappers now do not acquire the `OpenXRAPI` instance as a constructor parameter, but it is passed to them upon `OpenXRAPI` initialization through the `OpenXRExtensionWrapper::initialize()` method. This - along with making `registered_extension_wrappers` a static member of `OpenXRAPI` - allows us to register extension wrappers before `OpenXRAPI` construction.

Regarding the timing of extension wrapper registration, in the current design this needs to happen in core level extension initialization. The reason for this is that `OpenXRAPI` is currently instantiated and initialized at server level module initialization, which precedes server level extension initialization, so unless the Godot core is extended with an additional initialization level for OpenXR extension wrappers, they need to be initialized at core level.

OpenXR types are mapped to Godot's type system in the following way: XR handles (e.g. `XrSession`) are mapped to `uint64_t`. The same is true for enums such as `XrResult`. For structs such as `XrEventDataBuffer`, these are passed as pointers with `GDNativePtr`/`GDNativeConstPtr` instances. Currently there is an issue with the binding generator that results in a build error when `GDNativePtr`/`GDNativeConstPtr` is used as return type, so for these cases (e.g. `get_instance_proc_addr`) the current workaround is to pass the pointer as a 64-bit unsigned integer.

An issue that came up with GDExtensions in Godot 4.x is that library resources do not get properly globalized on Android: GDExtension libraries [can only be specified with resource paths](https://github.com/godotengine/godot/blob/master/editor/plugins/gdextension_export_plugin.h#L76), and the libraries get properly placed in the lib folder of the apk, but when loading the library, Godot tries to find it relative to the given resource path. To solve this issue, I modified `ProjectSettings::globalize_path()` so that on Android, libraries are loaded without their path specified. This fixes the issue of loading GDExtension libraries, although this might be considered a workaround, and further discussion might be needed on better ways to solve this issue.
```
+#ifdef ANDROID_ENABLED
+               if (p_path.ends_with(".so")) {
+                       return p_path.get_file();
+               }
+#endif
```

The PR also adds a new `"xr_features/use_anchor_api"` export setting, which is required for the implementation of extension wrappers for FB spatial anchors and scene capture.

Developed by [Migeran](https://migeran.com)

m4gr3d commented on core/config/project_settings.cpp:
```diff
@@ -256,6 +256,11 @@ bool ProjectSettings::get_ignore_value_in_docs(const String &p_name) const {
 
 String ProjectSettings::globalize_path(const String &p_path) const {
 	if (p_path.begins_with("res://")) {
+#ifdef ANDROID_ENABLED
+		if (p_path.ends_with(".so")) {
+			return p_path.get_file();
+		}
+#endif
```
Instead of making this change (which could have further unintended ramifications), let's update https://github.com/godotengine/godot/blob/master/editor/plugins/gdextension_export_plugin.h#L76 so that on Android (and iOS?), it's not restricted to the `res://` directory.

BastiaanOlij replied:
@konczg and a reminder to have a look at this.

konczg replied:
@m4gr3d @BastiaanOlij @kisg A possible issue with this approach could be that the exporting of these libraries is based on the paths specified in the gdextension configuration, which is the same path used for loading the library, so we cannot specify it as a path relative to the apk's lib folder, or else the exporting will be unsuccessful.

I modified the initial approach to resolve the library path in the GDExtensionResourceLoader to eliminate such unintended ramifications: https://github.com/godotengine/godot/pull/68259/files#diff-990645d1851b240a11471d20d4fda50e90b5d72a333565a5036b54807720b7a2R540
Could you please take a look at it?

m4gr3d replied:
@konczg The update looks good!

m4gr3d commented on modules/openxr/openxr_api.cpp:
```diff
@@ -1854,9 +1840,6 @@ OpenXRAPI::~OpenXRAPI() {
 	composition_layer_providers.clear();
 
 	// cleanup our extension wrappers
-	for (OpenXRExtensionWrapper *extension_wrapper : registered_extension_wrappers) {
-		memdelete(extension_wrapper);
-	}
```
Do we have any means of cleaning the extension wrappers?

BastiaanOlij replied:
This needs to move into `register_types.cpp/uninitialize_openxr_module` IMHO

BastiaanOlij replied:
@konczg just a reminder, this would be nice to add into the `uninitialize_openxr_module` function

konczg replied:
This has been added with @BastiaanOlij 's changes: https://github.com/godotengine/godot/pull/68259/files#diff-29befd060761763bf49d5349defccbc36170ce85ce687d92743b312acc330cf4R191

m4gr3d commented:
@BastiaanOlij We should also determine the criteria for an extension for 'graduating' from a GDExtension to being a core. 
The simplest approach would be to follow the OpenXR process, i.e: when a vendor extension moves to the core OpenXR spec, so should its extension wrapper.

BastiaanOlij commented:
> @BastiaanOlij We should also determine the criteria for an extension for 'graduating' from a GDExtension to being a core. The simplest approach would be to follow the OpenXR process, i.e: when a vendor extension moves to the core OpenXR spec, so should its extension wrapper.

Agreed, if it's in the OpenXR official headers and it doesn't clash with an existing core API, it can be in Godot itself. If it requires APIs that still have an unclear IP, are not open sourced or require a vendors SDK, it should be an external.

BastiaanOlij commented:
Finally had time to give this a proper look at. For the most part I fully agree with the changes done here, this seems like a sound approach.

The only thing I have serious questions about is `OpenXRAPIExtension` and how this is designed to work, I probably need to look at an actual extension implementation but at first glance, I think this isn't right.

There are a lot of things exposed that are already accessible by simply using OpenXRInterface that can be obtained by calling `XRServer::find_interface`or `XRServer::get_primary_interface`.
I also feel that some of these functions (like 'get_instance' and 'get_system_id') should be moved into `OpenXRExtensionWrapperExtension` and just added as public functions of this class. This class is already provided with a pointer to the OpenXRAPI class and it keeps things nicely contained instead of introducing a new class.

All the action stuff should be removed, this makes no sense, this is not how OpenXR is supposed to work. The actions and interaction profiles come from the action map the user define, it is the developer that determines which actions they want and how they want them setup, plugins should not mess with this.
The only two things that we may need to add here are:
- registering things into the metadata database (`openxr_defs.h/cpp`) which admittedly will need to be enhanced so plugins can register possible paths that can be used. 
- allowing plugins to add entries into the default action map that is created, though I'm not sure about this one as it potentially pollutes the action map with unnecessary actions.

Simply put, a developer should be able to trash the entire action map and build their own, as detailed or empty as they want, in fact this is recommended. The default action map is only there to create an example and a base and I even wonder if we haven't already put too much in there.

Remember, the actions the developer creates are the actions that are exposed to any rebinding UI and they need to make sense to the end user. If there are suddenly half a dozen entries that plugins start adding that overlap or duplicate entries that actually apply to the game, that is not a good user experience.

For instance, the VR editor currently runs on 4 actions mapped to all available devices. It has none of the extra entries that do not apply to the editor. Possibly a few more actions will be added, but they will be named and mapped appropriately. I expect, and indeed will teach, developers to be as succinct in creating their action maps.

I'm fully aware that there currently aren't any rebinding UIs (which is causing a lot of confusion in developers and a lot of game engines coming up with creative ways to work around this shortcoming) and that on Quest there is far less need for rebinding as there is only one input device, but we've got a UI that allows us to do the right thing.

BastiaanOlij commented:
Actually, thinking this through we have a bigger issue with the action map though the changes here around how we register extensions is going to be vital.

In the normal editor we never create an instance of the `OpenXRAPI`. We only do this when XR is enabled. This means we don't currently have a mechanism for registering ActionMap meta data with `OpenXRDefs` (which i plan to rename `OpenXRActionMapMetaData` which is far more fitting).

konczg commented:
@BastiaanOlij @kisg The current strategy was to expose to OpenXRExtensionWrapperExtension all parts of OpenXRAPI that OpenXRExtensionWrapper has access to, which is also the reason for putting them in a distinct OpenXRAPIExtension class, as I did not want to pollute OpenXRExtensionWrapperExtension with this many additional methods. After looking at it more thoroughly, it is clear that most of these should not be accessed by extension wrappers. However, isn't the problem here that the same OpenXRAPI is accessible both by OpenXRInterface and the extension wrappers, so even for the core extensions there is no separation of the interfaces? I think we should define more clearly which methods should be accessed by OpenXRInterface, and which by the extension wrappers.
Anyways, I listed the methods that I think should be kept accessible by OpenXRExtensionWrapperExtensions, are there any methods that we might not require here?
```
uint64_t get_instance();
uint64_t get_system_id();
uint64_t get_session();

Transform3D transform_from_pose(GDNativeConstPtr<const void> p_pose);
bool xr_result(uint64_t result, String format, Array args = Array());

static bool openxr_is_enabled(bool p_check_run_in_editor = true);
uint64_t get_instance_proc_addr(String p_name);
String get_error_string(uint64_t result);
String get_swapchain_format_name(int64_t p_swapchain_format);
bool is_initialized();
bool is_running();
uint64_t get_play_space();
int64_t get_next_frame_time();
bool can_render();
```

BastiaanOlij commented:
> @BastiaanOlij @kisg The current strategy was to expose to OpenXRExtensionWrapperExtension all parts of OpenXRAPI that OpenXRExtensionWrapper has access to, which is also the reason for putting them in a distinct OpenXRAPIExtension class, as I did not want to pollute OpenXRExtensionWrapperExtension with this many additional methods. After looking at it more thoroughly, it is clear that most of these should not be accessed by extension wrappers. However, isn't the problem here that the same OpenXRAPI is accessible both by OpenXRInterface and the extension wrappers, so even for the core extensions there is no separation of the interfaces? I think we should define more clearly which methods should be accessed by OpenXRInterface, and which by the extension wrappers. Anyways, I listed the methods that I think should be kept accessible by OpenXRExtensionWrapperExtensions, are there any methods that we might not require here?
> 
> ```
> uint64_t get_instance();
> uint64_t get_system_id();
> uint64_t get_session();
> 
> Transform3D transform_from_pose(GDNativeConstPtr<const void> p_pose);
> bool xr_result(uint64_t result, String format, Array args = Array());
> 
> static bool openxr_is_enabled(bool p_check_run_in_editor = true);
> uint64_t get_instance_proc_addr(String p_name);
> String get_error_string(uint64_t result);
> String get_swapchain_format_name(int64_t p_swapchain_format);
> bool is_initialized();
> bool is_running();
> uint64_t get_play_space();
> int64_t get_next_frame_time();
> bool can_render();
> ```

Ah that makes a lot more sense now that I understand to context.

The problem that we have is that the OpenXRAPI class contains a large part of the API to OpenXR, only some of it should be outward facing or accessible to extensions, most of it is consumed by OpenXRInterface only.

It's really where GDExtension fails a bit imho, for it to access stuff, we need to expose them to parts of the system where we don't really want it exposed.

That said, this makes a good argument for not introducing `OpenXRAPIExtension` but elevating `OpenXRAPI` to a proper Godot object type and then just being conservative with what we expose through `_bind_methods`.
That will probably have some consequences on how we instance and manage it, a singleton pattern might not be the best approach in this case, I'll have to think about that some more.

konczg commented:
@BastiaanOlij @kisg 

> The only thing I have serious questions about is `OpenXRAPIExtension` and how this is designed to work, I probably need to look at an actual extension implementation but at first glance, I think this isn't right.

We've published an example of implementing OpenXR extension wrappers through GDExtensions, it demonstrates how a wrapper for XR_FB_scene_capture can be added as an addon:
https://github.com/migeran/godot_openxr_meta

BastiaanOlij commented:
Sorry @konczg , took me awhile before I could look at this again. I think this is very close to what it should be. I'm still going back and forth whether we should just change `OpenXRApi` to a proper object but it makes a lot of sense to have the extension object to just expose those bits that are safe to expose. With it now cleaned up it's pretty straight forward.

@Faless if you haven't looked at this yet, can you have a look as well? Especially seeing konczg is still having issues with pointers. 

Also needs a rebase :)

BastiaanOlij commented on platform/android/export/export_plugin.cpp:
```diff
@@ -787,6 +787,12 @@ void EditorExportPlatformAndroid::_get_permissions(const Ref<EditorExportPreset>
 				r_permissions.push_back("com.oculus.permission.HAND_TRACKING");
 			}
 		}
+		bool use_anchor_api = p_preset->get("xr_features/use_anchor_api");
```
Isn't this something we can simply have in the plugin? Assuming it is build as an AAR it's able to add to the manifest if I'm not mistaken, and the inclusion of the plugin would require this to be true.

m4gr3d replied:
@konczg Take a look at https://github.com/godotengine/godot/pull/78958; in 4.2, we are expanding the capabilites of the android plugin to allow moving these type of configuration out of core.
You can see what the refactor looks like in https://github.com/GodotVR/godot_openxr_loaders/pull/38

So you should move these configs, as well as the anchor gdextension implementation in the Godot Openxr loader plugin instead of including them in core.

lyuma commented:
I'm concerned about the use of Multiple Inheritance with `class OpenXRExtensionWrapperExtension : public Object, OpenXRExtensionWrapper` given that `OpenXRExtensionWrapper` has member variables

one of the principles I've usually followed is to avoid MI in c++ unless all but the first parent have no members
In this case, `OpenXRExtensionWrapper` defines
```
        OpenXRAPI *openxr_api = nullptr;
        HashMap<String, bool *> request_extensions;
```
it's such a small amount of data that I think these belong in each of the child classes, and those should be the ones to implement the `get_request_extensions()` function and the constructor
That way, `OpenXRExtensionWrapper` can be a class with only virtual functions. perhaps `get_request_extensions()` can be declared pure virtual so implementers are forced to store the `request_extensions`

BastiaanOlij commented:
Hmm weird @lyuma , I thought I had reacted to this but I guess I either didn't post my reaction or lost it somehow.

This approach kinda grew out of how things were in the plugin so I think we can clean this up nicely.

We can get rid of the `openxr_api` member variable all together. We can just use `OpenXRAPI::get_singleton()`. This also works nicer with when GDExtension is used as the wrapper is then used. 

I think we can also get rid of `request_extensions`. We can simply implement  `HashMap<String, bool *> get_request_extensions()` on each extension and return the map, there is no need to keep this as a member variable.

I'm on the fence about whether we should stick with MI, I don't really see a point to make all the internal wrapper objects proper Godot objects because it just adds a bunch of objects not intended for end users to pickup. I guess we can only register the virtual classes and GDExtension wrapper class, but just using MI on the GDExtension wrapper seems fine to me.

konczg committed:
Add GDExtension support for OpenXR extension wrappers

This commit adds the classes OpenXRExtensionWrapperExtension and OpenXRAPIExtension
that can be used in GDExtensions to define OpenXR extension wrappers.
It modifies extension wrapper registration so that they can be registered
before OpenXRAPI instantiation (e.g. in core level initialization of GDExtensions).

Developed by Migeran (https://migeran.com)

BastiaanOlij commented:
@konczg note that I've put some overlapping changes into #70694 including applying the changes requested by @lyuma .
Once merged you can add your two `GDExtension` classes on top of that work.

konczg commented:
@BastiaanOlij I have rebased the changes here to https://github.com/BastiaanOlij/godot/tree/change_openxr_extensions . Once this is merged I can further rebase to main if it will be needed. I also updated https://github.com/migeran/godot_openxr_meta to reflect the changes.

kisg commented:
Hi @BastiaanOlij,

apart from rebasing this to the latest master are there any issues that need to be solved before it can be merged? It would be great to get this one into 4.0...

BastiaanOlij commented on doc/classes/OpenXRAPIExtension.xml:
```diff
@@ -0,0 +1,89 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="OpenXRAPIExtension" inherits="RefCounted" version="4.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../class.xsd">
+	<brief_description>
+	</brief_description>
```
It would be good to add in description in the XML files

BastiaanOlij commented on core/extension/gdextension.cpp:
```diff
@@ -537,7 +537,13 @@ Ref<Resource> GDExtensionResourceLoader::load(const String &p_path, const String
+#ifdef ANDROID_ENABLED
+	// On Android, the libraries are placed in the lib folder of the apk,
+	// so we only use relative paths here.
+	String abs_path = library_path.get_file();
```
This would be good to get confirmation from the android team for, but I guess this makes sense

m4gr3d replied:
To add to this, this logic makes sense for gdextension packaged within Android apks. 

The logic will need to be updated for the Godot Android Editor since it should be able to load gdextension plugins outside of its own apk. I can take a look in making it work after this pr is merged.

m4gr3d replied:
@dsnopek FYI, this means that gdextension has been broken on Android up to now.
This is something we may want to include in the Godot 4.2 release notes.

BastiaanOlij commented on modules/openxr/extensions/openxr_extension_wrapper_extension.cpp:
```diff
@@ -0,0 +1,206 @@
+HashMap<String, bool *> OpenXRExtensionWrapperExtension::get_request_extensions() {
+	Dictionary request_extension;
+
+	if (GDVIRTUAL_CALL(_get_request_extensions, request_extension)) {
```
This is more a reminder for later, possibly 4.1, but in hindsight I think parsing a pointer to a bool around isn't really suitable for GDExtensions. We did it this way in the plugin in Godot 3 so I kept that approach, alas.

We might change this to work similar to what we've done on the Vulkan side where we keep everything name based on the extension. So we simply record we want an extension enabled, and whether it is a mandatory extension, and have a separate call that checks whether and extension was enabled.

But I think that is too late now to change so close to 4.0 release.

BastiaanOlij reviewed (approved):
Once rebased so it is just the last commit ontop of the latest master, and documentation is completed, I think this is good to go.

My further remarks are mostly just to keep in mind for later, not show stoppers (plus one that requires a rethink not directly related to this PR but choices made earlier).

m4gr3d commented:
@BastiaanOlij @konczg Was there anything that remained to be done prior to merging this PR?

BastiaanOlij commented:
@m4gr3d wow blast from the past. it'll need to be rebased again but I think all my issues with it were resolved. We never adjusted how we register extensions, I only did that work on the Vulkan drivers so thats something for a later date, maybe even v5 by now.

It would still be nice to fill in the documentation files but that too can be done in a follow up PR.

So if you don't see anything that you're worried about, I suggest we do merge this. It adds the ability to access stuff from GDExtension so I don't think this has any risks of breaking stuff.

konczg commented:
@m4gr3d @BastiaanOlij I will try to rebase this tomorrow.

konczg commented:
@BastiaanOlij @m4gr3d @kisg I've rebased this to master. The [MethodBind ValidatedCall optimization](https://github.com/godotengine/godot/commit/1c93606e470f0cad5f14af104ccb89a95a89931c) broke the usage of `GDExtensionPtr`/`GDExtensionConstPtr` in GDNative, which is also fixed in this PR [here](https://github.com/godotengine/godot/pull/68259/files#diff-dd63ab4bde66568e0fc386d410467657636606b08fb049073013a748d0db75f9R56).
The rebased version was tested on Oculus Quest 2 with the following OpenXR Extension Wrapper: https://github.com/migeran/godot_openxr_meta

BastiaanOlij reviewed (approved):
lgtm, since we're in feature freeze this probably will need to move to 4.2, but I believe this can be merged as soon as 4.1 has been released.

kisg commented:
> lgtm, since we're in feature freeze this probably will need to move to 4.2, but I believe this can be merged as soon as 4.1 has been released.

@BastiaanOlij Oh, this was not clear to us, that this could only go into 4.2 (it was not mentioned in the last few comments between you and @m4gr3d, or when @konczg said that he would rebase it). Then we most likely would not have rushed this rebase, because it will probably have to be rebased again after the 4.1 release. :(

Isn't it possible to merge it before the next beta? I think this is a low-risk change + it also includes a GDExtension bug fix, which should definitely go into 4.1, as @konczg mentioned: https://github.com/godotengine/godot/pull/68259/files#diff-dd63ab4bde66568e0fc386d410467657636606b08fb049073013a748d0db75f9R56

m4gr3d commented:
@kisg @konczg Can you pull the bug fix into a separate PR; that would be okay to merge in `4.1` since the feature freeze / beta period is to address and incorporate such fixes. cc @dsnopek for a look at the GDExtension fix.

> it will probably have to be rebased again after the 4.1 release

I don't expect a significant rebase for `4.2`; since `4.1` is in feature freeze, it's unlikely for the codebase to change much between now and when `4.1` stable is released.

Targeting `4.2` may actually be beneficial; this PR significantly improves the engine's ability to integrate and use vendor's extensions, so we may want to mark it as a key (new) feature for `4.2` with corresponding blog post, documentation and samples.
One area that still needs fleshing out is the delivery mechanism for OpenXR capabilities provided through GDExtension. Your example project includes all the changes (project and extensions); at least on Quest we'll want to leverage Godot Android plugins to easily provide the capability.

BastiaanOlij commented:
Indeed, it's unlikely a rebase for 4.2 will be a big thing, basically we're in bug fixing mode only. I agree with @m4gr3d that it would be a nice feature to announce for 4.2

@kisg @konczg in case you missed the announcement, Godot is attempting to run on a 3 month release schedule. 2 months of development, 1 month of testing and fixing. We're now in the month of fixing with a release planned early July.
Then July + August will be the merge window for development in 4.2, with bug fixing in September and release in Early October.
Then October + November merge window for development in 4.3, bug fixing in December, release Early Jan, etc.

So if we're happy with the work here and Fredia and me approve this, we should be able to get it merge it as is as soon as 4.1 is released.

akien-mga commented:
> Godot is attempting to run on a 3 month release schedule. 2 months of development, 1 month of testing and fixing.

Just to clarify, it's a 4-month cycle (3 months merging features, 1 month feature freeze).

konczg commented:
@BastiaanOlij @m4gr3d @kisg I've rebased this to master.

BastiaanOlij commented:
The team wants to do a 4.1.1 patch at the end of the week before we start merging, hopefully soon after this can be one of the first PRs merged for 4.2-dev :)

YuriSizov commented:
@konczg Could you please squash your commits into one? Make sure that the final commit has a short but descriptive message (the title of this PR is a good option). See [this documentation](https://docs.godotengine.org/en/stable/contributing/workflow/pr_workflow.html#the-interactive-rebase), if you need help with squashing.

As mentioned during the review, it would be great to add at least some descriptions to the newly created documentation. At the very least we need class descriptions, but it would also be amazing if you could fill in method and property description as well.

YuriSizov commented:
@konczg I'm not sure if this was done on purpose or not, but currently your commit is attributed to Bastiaan, while you are only tagged as a committed. You should probably amend the author information, or at least add yourself as a co-author using this construct at the end of your commit message body

```
Co-authored-by: Your Name <your_email>
```

konczg commented:
@YuriSizov Sorry, this was done by accident, I've fixed the commit.

m4gr3d commented on core/extension/gdextension.cpp:
```diff
@@ -635,7 +635,13 @@ Ref<Resource> GDExtensionResourceLoader::load(const String &p_path, const String
 
 	Ref<GDExtension> lib;
 	lib.instantiate();
+#ifdef ANDROID_ENABLED
+	// On Android, the libraries are placed in the lib folder of the apk,
+	// so we only use relative paths here.
+	String abs_path = library_path.get_file();
```
@konczg Looks like this was fixed already in https://github.com/godotengine/godot/pull/68362, so we should be able to revert this change.

m4gr3d commented on platform/android/export/export_plugin.cpp:
```diff
@@ -836,6 +836,12 @@ void EditorExportPlatformAndroid::_get_permissions(const Ref<EditorExportPreset>
 				r_permissions.push_back("com.oculus.permission.HAND_TRACKING");
 			}
 		}
+		bool use_anchor_api = p_preset->get("xr_features/use_anchor_api");
```
@konczg Can you move these configs to the Godot OpenXR loader plugin as mentioned in https://github.com/godotengine/godot/pull/68259/files#r1265943701

m4gr3d replied:
Thanks for the update! Let me know if you need assistance with updating the Godot OpenXR loader plugin with the anchor configs.

m4gr3d reviewed (changes requested):
Thanks for the quick update!

Can you clean up the anchor options as mentioned in the comments.

m4gr3d reviewed (approved):
Looks great!

BastiaanOlij commented:
@YuriSizov is there any input the production team would like to give here? It would be nice to get this merged :)

YuriSizov commented:
> is there any input the production team would like to give here?

Not sure yet, but we have it marked for dev2, so the aim would be to finalize and merge it before that. 👍

YuriSizov commented on modules/openxr/extensions/openxr_extension_wrapper_extension.cpp:
```diff
@@ -0,0 +1,206 @@
+/**************************************************************************/
+
+#include "openxr_extension_wrapper_extension.h"
+#include "../openxr_api.h"
```
Needs an empty line in between.

YuriSizov commented on modules/openxr/action_map/openxr_interaction_profile_meta_data.cpp:
```diff
@@ -29,7 +29,6 @@
 /**************************************************************************/
 
 #include "openxr_interaction_profile_meta_data.h"
-
```
Shouldn't remove it the line here.

konczg replied:
@YuriSizov I've updated the code

YuriSizov commented on doc/classes/OpenXRExtensionWrapperExtension.xml:
```diff
@@ -0,0 +1,175 @@
+		<method name="_on_state_visible" qualifiers="virtual">
+			<return type="void" />
+			<description>
+				Called when the OpenXR session state is changed to visible, OpenXR is now ready to receive frames.
```
```suggestion
				Called when the OpenXR session state is changed to visible. This means OpenXR is now ready to receive frames.
```

YuriSizov commented on doc/classes/OpenXRExtensionWrapperExtension.xml:
```diff
@@ -0,0 +1,175 @@
+		<method name="_on_state_synchronized" qualifiers="virtual">
+			<return type="void" />
+			<description>
+				Called when the OpenXR session state is changed to synchronized.
+				OpenXR also returns to this state when the application looses focus.
```
```suggestion
				Called when the OpenXR session state is changed to synchronized. OpenXR also returns to this state when the application loses focus.
```

You need to avoid new lines in the docs, unless you want to create a new paragraph (because that's what a new line here does). In this case it should be the same paragraph, as the sentences are short and the context is the same.

Also a typo.

YuriSizov commented on doc/classes/OpenXRExtensionWrapperExtension.xml:
```diff
@@ -0,0 +1,175 @@
+		<method name="_on_state_focused" qualifiers="virtual">
+			<return type="void" />
+			<description>
+				Called when the OpenXR session state is changed to focused.
+				This state is the active state when the game runs.
```
```suggestion
				Called when the OpenXR session state is changed to focused. This state is the active state when the game runs.
```

YuriSizov commented on doc/classes/OpenXRExtensionWrapperExtension.xml:
```diff
@@ -0,0 +1,175 @@
+			<return type="void" />
+			<param index="0" name="session" type="int" />
+			<description>
+				Called right after we've successfully created the OpenXR session.
```
```suggestion
				Called right after the OpenXR session is created.
```

YuriSizov commented on doc/classes/OpenXRExtensionWrapperExtension.xml:
```diff
@@ -0,0 +1,175 @@
+		<method name="_on_register_metadata" qualifiers="virtual">
+			<return type="void" />
+			<description>
+				Allows extensions to register additional controller metadata.
+				This function is called even when [code]OpenXRAPI[/code] is not constructured as the metadata needs to be available to the editor.
+				Extensions should also provide metadata regardless of whether they are supported on the host system as the controller data is used to setup action maps for users who may have access to the relevant hardware.
```
```suggestion
				Allows extensions to register additional controller metadata. This function is called even when [code]OpenXRAPI[/code] is not constructed as the metadata needs to be available to the editor.
				Extensions should also provide metadata regardless of whether they are supported on the host system. The controller data is used to setup action maps for users who may have access to the relevant hardware.
```

What is `OpenXRAPI` referring to here? If it's a part of the documented API, it should be cross-linked. And if it's not, it should probably be written without the code formatting, as plain text.

Other than that, there is the same issue with line breaks as before, a typo, and also an overly long sentences. Try to split up sentences in the documentation, as long, complexly connected descriptions are harder to comprehend.

YuriSizov commented on doc/classes/OpenXRExtensionWrapperExtension.xml:
```diff
@@ -0,0 +1,175 @@
+		<method name="_on_process" qualifiers="virtual">
+			<return type="void" />
+			<description>
+				Called as part of the OpenXR process handling.
+				This happens right before physics process and normal processing is run.
+				This is when controller data is queried and made available to game logic.
```
```suggestion
				Called as part of the OpenXR process handling. This happens right before general and physics processing steps of the main loop. During this step controller data is queried and made available to game logic.
```

YuriSizov commented on doc/classes/OpenXRExtensionWrapperExtension.xml:
```diff
@@ -0,0 +1,175 @@
+		<method name="_on_pre_render" qualifiers="virtual">
+			<return type="void" />
+			<description>
+				Called right before we start rendering the XR viewports.
```
```suggestion
				Called right before the XR viewports begin their rendering step.
```

YuriSizov commented on doc/classes/OpenXRExtensionWrapperExtension.xml:
```diff
@@ -0,0 +1,175 @@
+			<return type="void" />
+			<param index="0" name="instance" type="int" />
+			<description>
+				Called right after we've successfully created the OpenXR instance.
```
```suggestion
				Called right after the OpenXR instance is created.
```

YuriSizov commented on doc/classes/OpenXRAPIExtension.xml:
```diff
@@ -0,0 +1,109 @@
+		<method name="is_running">
+			<return type="bool" />
+			<description>
+				Returns [code]true[/code] if OpenXR is running ([code]xrBeginSession[/code] was successfully called and the swapchains were created).
```
```suggestion
				Returns [code]true[/code] if OpenXR is running (i.e. [code]XrBeginSession[/code] was successfully called and the swapchains were created).
```

I assume it needs to be capitalized like all other `Xr` codes?

konczg replied:
@YuriSizov  [xrBeginSession](https://registry.khronos.org/OpenXR/specs/1.0/man/html/xrBeginSession.html) is an OpenXR function, so by the used naming convention it shouldn't be capitalized.

YuriSizov commented on doc/classes/OpenXRAPIExtension.xml:
```diff
@@ -0,0 +1,109 @@
+			<return type="int" />
+			<param index="0" name="name" type="String" />
+			<description>
+				Returns the function pointer of the OpenXR function with the specified name, cast to an integer.
+				[code]openxr/util.h[/code] contains macros for ease of use in acquiring OpenXR functions, e.g. [code]GDEXTENSION_INIT_XR_FUNC_V(xrCreateAction)[/code].
+				If the function with the given name does not exist, the method returns [code]0[/code].
```
```suggestion
				Returns the function pointer of the OpenXR function with the specified name, cast to an integer. If the function with the given name does not exist, the method returns [code]0[/code].
				[b]Note:[/b] [code]openxr/util.h[/code] contains utility macros for acquiring OpenXR functions, e.g. [code]GDEXTENSION_INIT_XR_FUNC_V(xrCreateAction)[/code].
```

konczg replied:
@YuriSizov I've updated the documentation accordingly

YuriSizov commented on doc/classes/OpenXRAPIExtension.xml:
```diff
@@ -0,0 +1,109 @@
+		<method name="get_instance">
+			<return type="int" />
+			<description>
+				Returns the [code]XrInstance[/code] craated at the initialization of [code]OpenXRAPI[/code].
```
```suggestion
				Returns the [code]XrInstance[/code] created during the initialization of [code]OpenXRAPI[/code].
```

Same question as in previous comments regarding `OpenXRAPI`.

konczg replied:
@YuriSizov `OpenXRAPI` is an internal Godot class that is used by core OpenXR extensions to access the OpenXR API (e.g. `OpenXROpenGLExtension`), so I don't think there is any documentation we could currently refer to here.

YuriSizov commented on doc/classes/OpenXRAPIExtension.xml:
```diff
@@ -0,0 +1,109 @@
+			<return type="String" />
+			<param index="0" name="result" type="int" />
+			<description>
+				Returns an error string for a given [code]XrResult[/code].
```
```suggestion
				Returns an error string for the given [code]XrResult[/code].
```

konczg replied:
@YuriSizov I've updated the documentation accordingly

YuriSizov reviewed (changes requested):
Some code style changes and documentation clarifications are required (as well as a few formatting and grammar issues).

konczg commented on doc/classes/OpenXRAPIExtension.xml:
```diff
@@ -0,0 +1,109 @@
+			<param index="2" name="args" type="Array" />
+			<description>
+				Returns [code]true[/code] if the provided [code]XrResult[code] (cast to an integer) is successful.
+				Otherwise returns [code]false[/code] and prints the [code]XrResult[/code] converted to a string, with the specified additional information.
```
@YuriSizov I've added the relevant OpenXR documentation pages both as description urls, and as tutorial links.

konczg commented on modules/openxr/openxr_api.cpp:
```diff
@@ -38,6 +38,8 @@
 #include "core/os/memory.h"
 #include "core/version.h"
 
+#include "extensions/openxr_extension_wrapper_extension.h"
```
@YuriSizov I've updated the code accordingly

konczg commented on modules/openxr/openxr_api_extension.h:
```diff
@@ -0,0 +1,75 @@
+#include "core/os/os.h"
+#include "core/os/thread_safe.h"
+#include "core/variant/native_ptr.h"
+#include "openxr_api.h"
```
@YuriSizov I've updated the code accordingly

konczg commented on modules/openxr/extensions/openxr_extension_wrapper_extension.h:
```diff
@@ -0,0 +1,114 @@
+#include "core/os/os.h"
+#include "core/os/thread_safe.h"
+#include "core/variant/native_ptr.h"
+#include "openxr_extension_wrapper.h"
```
@YuriSizov I've updated the code

YuriSizov commented:
@konczg Thanks for the updates! There were a few more notes that you've probably missed because GitHub collapsed them. :) By the way, you don't need to reply to every one of them, you can just hit the "Resolve conversation" button to mark it as resolved.

Regarding `OpenXRAPI`, if it's not some object that we can refer users to, then we should probably just write it as plain text, "OpenXR API", with no formatting.

AThousandShips commented on modules/openxr/openxr_api_extension.h:
```diff
@@ -0,0 +1,76 @@
+	uint64_t get_system_id();
+	uint64_t get_session();
+
+	// helper method to convert an XrPosef to a Transform3D
```
```suggestion
	// Helper method to convert an XrPosef to a Transform3D.
```
As per the [style guide](https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html#comment-style-guide)

YuriSizov reviewed (approved):
I think docs look good now, thanks!

YuriSizov commented:
Thanks!
